zcash_transparent/src/bundle.rs

```diff
@@ -258,6 +267,72 @@ impl TxIn<Authorized> {
     }
 }
 
+impl TxIn<builder::Coinbase> {
+    /// Creates an input for a coinbase transaction. Does not support creating inputs for the
+    /// genesis block.
+    pub fn coinbase(height: BlockHeight, miner_data: &MinerData) -> Result<Self, coinbase::Error> {
+        let mut script_sig = match i64::from(height) {
+            0 => Err(coinbase::Error::GenesisInputNotSupported)?,
+            h => push_num(h).to_bytes(),
+        };
+
+        // # Consensus
+        //
+        // > A coinbase transaction for a block at block height greater than 0 MUST have a script
+        // > that, as its first item, encodes the block height `height` as follows. For `height` in
+        // > the range {1 .. 16}, the encoding is a single byte of value `0x50` + `height`.
+        // > Otherwise, let `heightBytes` be the signed little-endian representation of `height`,
+        // > using the minimum nonzero number of bytes such that the most significant byte is <
+        // > `0x80`. The length of `heightBytes` MUST be in the range {1 .. 5}. Then the encoding is
+        // > the length of `heightBytes` encoded as one byte, followed by `heightBytes` itself. This
+        // > matches the encoding used by Bitcoin in the implementation of [BIP-34] (but the
+        // > description here is to be considered normative).
+        //
+        // <https://zips.z.cash/protocol/protocol.pdf#txnconsensus>
+        //
+        // ## Note
+        //
+        // The height is encoded in the `push_num` fn above.
+        //
+        // [BIP-34]: <https://github.com/bitcoin/bips/blob/master/bip-0034.mediawiki>
+        if script_sig.is_empty() {
+            Err(coinbase::Error::EmptyHeight)?;
+        } else if script_sig.len() > coinbase::MAX_COINBASE_HEIGHT_LEN {
```

Comment:
This error cannot occur. `push_num` always returns an opcode, and every opcode serializes to at least one byte. We can therefore check this with an assertion, as a failure here would be a programming error.
```suggestion
        assert!(!script_sig.is_empty());
        if script_sig.len() > coinbase::MAX_COINBASE_HEIGHT_LEN {
```